packages/widget-v2/src/pages/SwapPage/useUpdateAmountWhenRouteChanges.ts

```diff
@@ -17,18 +18,24 @@ export const useUpdateAmountWhenRouteChanges = () => {
 
   useEffect(() => {
     if (!route.data || !sourceAsset || !destinationAsset) return;
+    if (sourceAsset?.amount === "" && direction === "swap-in") return;
+    if (destinationAsset?.amount === "" && direction === "swap-out") return;
 
     const swapInAmount = convertTokenAmountToHumanReadableAmount(
       route.data.amountOut,
-      destinationAsset.decimals
+      DEFAULT_DECIMAL_PLACES
     );
     const swapOutAmount = convertTokenAmountToHumanReadableAmount(
       route.data.amountIn,
-      sourceAsset.decimals
+      DEFAULT_DECIMAL_PLACES
     );
 
-    const swapInAmountChanged = swapInAmount !== destinationAsset.amount;
-    const swapOutAmountChanged = swapOutAmount !== sourceAsset.amount;
+    const swapInAmountChanged =
+      removeTrailingZeros(swapInAmount) !==
+      removeTrailingZeros(destinationAsset.amount);
+    const swapOutAmountChanged =
+      removeTrailingZeros(swapOutAmount) !==
+      removeTrailingZeros(sourceAsset.amount);
```

Comment:
using the `DEFAULT_DECIMAL_PLACES` instead of the native decimals and using `removeTrailingZeros` before comparing the two is what fixes the infinite set state